reference/tests/AddOpTest.cpp

```diff
@@ -0,0 +1,290 @@
+  constexpr llvm::StringRef kModule = R"(
+  module {
+    func.func @main(%arg0: tensor<2x5xf32>, %arg1: tensor<2x5xf32>) -> tensor<2x5xf32> {
+      %result = stablehlo.add %arg0, %arg1 : tensor<2x5xf32>
+      func.return %result : tensor<2x5xf32>
+    }
+  })";
+  runTestCase(kModule, {/*operands*/
+                        {"0.0", "-0.0", "1.0", "0.125", "0.1", "3.14159265",
+                         "inf", "inf", "-inf", "inf"},
+                        {"0.0", "-0.0", "7.0", "0.75", "0.3", "3.14159265", "0",
+                         "inf", "-inf", "-inf"},
+                        /*expected result*/
+                        {"0.0", "-0.0", "8.0", "0.875", "0.4", "6.2831855",
+                         "inf", "inf", "-inf", "nan"}});
+}
+
+TEST(AddOpInterpreterTest, BF16) {
+  constexpr llvm::StringRef kModule = R"(
+  module {
+    func.func @main(%arg0: tensor<2x5xbf16>, %arg1: tensor<2x5xbf16>) -> tensor<2x5xbf16> {
+      %result = stablehlo.add %arg0, %arg1 : tensor<2x5xbf16>
+      func.return %result : tensor<2x5xbf16>
+    }
+  })";
+  runTestCase(kModule, {/*operands*/
+                        {"0.0", "-0.0", "1.0", "0.125", "0.1", "3.140625",
+                         "inf", "inf", "-inf", "inf"},
+                        {"0.0", "-0.0", "7.0", "0.75", "0.3", "3.140625", "0",
+                         "inf", "-inf", "-inf"},
+                        /*expected result*/
+                        {"0.0", "-0.0", "8.0", "0.875", "0.4", "6.28125", "inf",
+                         "inf", "-inf", "nan"}});
+}
+
+TEST(AddOpInterpreterTest, F64) {
+  constexpr llvm::StringRef kModule = R"(
+  module {
+    func.func @main(%arg0: tensor<2x5xf64>, %arg1: tensor<2x5xf64>) -> tensor<2x5xf64> {
+      %result = stablehlo.add %arg0, %arg1 : tensor<2x5xf64>
+      func.return %result : tensor<2x5xf64>
+    }
+  })";
+
+  runTestCase(kModule, {/*operands*/
+                        {"0.0", "-0.0", "1.0", "0.125", "0.1",
+                         "3.14159265358979323846", "inf", "inf", "-inf", "inf"},
+                        {"0.0", "-0.0", "7.0", "0.75", "0.3",
+                         "3.14159265358979323846", "0", "inf", "-inf", "-inf"},
+                        /*expected result*/
+                        {"0.0", "-0.0", "8.0", "0.875", "0.4",
+                         "6.283185307179586", "inf", "inf", "-inf", "nan"}});
+}
+
+TEST(AddOpInterpreterTest, SInt4WithOv) {
+  constexpr llvm::StringRef kModule = R"(
+  module {
+    func.func @main(%arg0: tensor<2x2x2xi4>, %arg1: tensor<2x2x2xi4>) -> tensor<2x2x2xi4> {
+      %result = stablehlo.add %arg0, %arg1 : tensor<2x2x2xi4>
+      func.return %result : tensor<2x2x2xi4>
+    }
+  })";
+  runTestCase(kModule, {/*operands*/
+                        {"0", "1", "2", "-3", "7", "-8", "7", "-8"},
+                        {"-8", "-1", "2", "-3", "1", "-1", "7", "-8"},
+                        /*expected result*/
+                        {"-8", "0", "4", "-6", "-8", "7", "-2", "0"}});
```

Comment:
The spec draft says that overflow is implementation-defined. Should we be testing it in that case?

Reply:
I agree. Also, given that, we have plans to include the tests link from here  in the spec, it does not make sense to include such cases.